File: backend/schemas/agent_terminal.py
"""Pydantic schemas for the embedded agent terminal API.

Mirrors :class:`backend.db.models.agent_terminal.AgentTerminalSession`
for REST responses, plus separate input schemas for spawn / WS messages.
"""

from __future__ import annotations

import re
import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

_SLUG_RE = re.compile(r"^[a-z][a-z0-9-]*[a-z0-9]$|^[a-z]$")

# CR-V2-007: SPAWN role — the interactive terminal is the AI Agent (the doer; replaces the v1
# Coordinator). Used by AgentTerminalSpawnRequest so the spawn API rejects every other role at the
# request boundary. These are CHARTER-PATH SLUGS (hyphen) — ``role`` flows straight to the filesystem
# charter path ``.claude/agents/<role>/CLAUDE.md``; the DB-side ``ai_agent`` (underscore) maps to it via
# orchestrator._charter_slug_for_role.
AgentRole = Literal["ai-agent"]
# Debug-attach (CR-NS-018 §10) targets an orchestrator-backed agent session — you attach to the AI Agent
# or the independent Auditor. A deliberately SEPARATE type from the spawn AgentRole (spawn ≠ debug-attach);
# this is also the READ type, because the SAME agent_terminal_sessions table holds debug-attach rows.
DebugAttachRole = Literal["ai-agent", "auditor"]
TerminatedBy = Literal["idle", "user", "crash", "server_restart"]


class AgentTerminalSessionRead(BaseModel):
    """Single agent terminal session row (active or historical)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    # READ accepts all debug-attach roles: this row may be an AI-Agent spawn OR an Auditor debug-attach
    # session (CR-V2-007). AI-Agent-only lives on the SpawnRequest, not here.
    role: DebugAttachRole
    project_slug: str
    pid: int
    created_at: datetime
    ended_at: Optional[datetime]
    exit_code: Optional[int]
    terminated_by: Optional[TerminatedBy]
    last_activity_at: datetime


class AgentTerminalSpawnRequest(BaseModel):
    """POST /agent-terminal/spawn body."""

    role: AgentRole
    project_slug: str = Field(min_length=1, max_length=100)

    @field_validator("project_slug")
    @classmethod
    def _validate_slug(cls, v: str) -> str:
        if not _SLUG_RE.match(v):
            raise ValueError(f"Invalid project slug: {v!r}")
        return v

File: frontend/src/services/api/agentTerminal.ts
/**
 * API client for ``/api/v1/agent-terminal/*``.
 *
 * Backend: :file:`backend/api/routes/agent_terminal.py`. Surfaces the
 * embedded PTY-backed claude CLI sessions — REST for lifecycle, WebSocket
 * for IO streaming. In v2.0.0 this raw PTY is the **break-glass debug**
 * path only (CR-V2-015): the first-class Manažér↔AI-Agent channel is the
 * event-rendered transcript + engine relay on the Riadiace centrum (spine
 * STEP 1), NOT this raw byte stream.
 *
 * All endpoints require the ``ri`` role; the auth wrapper in ``api.ts``
 * surfaces 403 as a thrown ``ApiError`` the page can render as a Lock
 * placeholder.
 */

import api from "../api";

// CR-V2-022 (OQ-7 follow-on): the v1 spawn role ``"coordinator"`` is re-keyed to the v2 AI Agent across the
// store/api/PersistentTerminalsLayer — the v2 vocabulary has ONE doer (the AI Agent), no Coordinator. The
// VALUE is the HYPHEN charter-path slug ``"ai-agent"`` to match the backend wire contract verbatim
// (``AgentTerminalSpawnRequest.role`` + ``AgentTerminalSessionRead.role`` are ``Literal["ai-agent"]``, the
// on-disk charter slug — CR-V2-007/015); keeping the same value is what "sweep without breaking sessions"
// requires, so a cold-start ``refresh()`` re-matches existing rows and a spawn POST stays valid. This is the
// SPAWN role for the warm `claude` session; the debug-attach (CR-V2-015 §10) targets a pipeline role via
// `DebugAttachRole` in `pipeline.ts` — a deliberately separate type (spawn ≠ debug-attach).
export type AgentRole = "ai-agent";

export type TerminatedBy = "idle" | "user" | "crash" | "server_restart";

export interface AgentTerminalSession {
  id: string;
  user_id: string;
  role: AgentRole;
  project_slug: string;
  pid: number;
  created_at: string;
  ended_at: string | null;
  exit_code: number | null;
  terminated_by: TerminatedBy | null;
  last_activity_at: string;
}

export interface SpawnRequest {
  role: AgentRole;
  project_slug: string;
}

/** Spawn a fresh claude CLI process for ``(role, project_slug)``. */
export function spawnAgentTerminalApi(
  payload: SpawnRequest,
): Promise<AgentTerminalSession> {
  return api.post<AgentTerminalSession>("/agent-terminal/spawn", payload);
}

/** List active (``ended_at IS NULL``) sessions for the current user. */
export function listAgentTerminalSessionsApi(): Promise<AgentTerminalSession[]> {
  return api.get<AgentTerminalSession[]>("/agent-terminal/sessions");
}

/** Per-role charter availability for a project (CR-NS-014). A role is
 *  available when its ``.claude/agents/<role>/CLAUDE.md`` exists. In v2 the
 *  only spawnable role is the AI Agent (CR-V2-007). */
export type AvailableRoles = Record<AgentRole, boolean>;

export function getAvailableRolesApi(projectSlug: string): Promise<AvailableRoles> {
  return api.get<AvailableRoles>("/agent-terminal/available-roles", {
    params: { project_slug: projectSlug },
  });
}

/** Explicit End session — SIGTERM, grace, SIGKILL. Idempotent. */
export function endAgentTerminalSessionApi(
  sessionId: string,
): Promise<void> {
  return api.delete<void>(`/agent-terminal/sessions/${sessionId}`);
}

/**
 * Build the WebSocket URL for a given session. The browser ``WebSocket``
 * constructor cannot set headers, so the JWT travels as a query-string
 * ``token`` parameter — the backend route ``WS /ws/{session_id}``
 * decodes it the same way as the REST ``Authorization: Bearer`` flow.
 *
 * Scheme is derived from the current page (``wss://`` if the page is
 * served over HTTPS, otherwise ``ws://``). API base path is fixed to
 * ``/api/v1`` mirroring ``api.ts``.
 */
export function buildAgentTerminalWsUrl(
  sessionId: string,
  token: string,
): string {
  const protocol = window.location.protocol === "https:" ? "wss:" : "ws:";
  const base =
    (import.meta.env.VITE_API_BASE_URL as string | undefined) ||
    `${protocol}//${window.location.host}`;
  // Strip http(s) → ws(s) if VITE_API_BASE_URL was an http URL.
  const wsBase = base.replace(/^http/, "ws");
  return `${wsBase}/api/v1/agent-terminal/ws/${sessionId}?token=${encodeURIComponent(token)}`;
}
